crates/flux-middle/src/fhir.rs

```diff
@@ -836,8 +839,8 @@ pub struct Defn {
 }
 
 impl Generics {
-    pub(crate) fn get_param(&self, def_id: LocalDefId) -> &GenericParam {
-        self.params.iter().find(|p| p.def_id == def_id).unwrap()
+    pub(crate) fn get_param(&self, def_id: LocalDefId) -> Option<&GenericParam> {
+        self.params.iter().find(|p| p.def_id == def_id)
```

Comment:
You are hiding a problem somewhere else. I think the problem is this test you added

https://github.com/flux-rs/flux/pull/589/files#diff-951d1aeaa0d86a85e279fc5169ff954af98a94aa53e0a6ca824b3c4d3c44ff7cR11-R18

If you specify an abstract refinement you need to specify all the generics of the function. The following works

```rust
#[flux::sig(fn<T, refine q: T -> bool>(&T{v:q(v)}) -> T{v: q(v)})]
```

The issue is that we don't check the refined signature has enough generics. Alternatively, we could lift the rust generics if the flux signature only has refinement parameters.

Reply:
I'm not following this comment. 

1. I put the `Option` in over there because otherwise we just get a strange crash,
2. (In `main`) if you just write the signature you wrote, you get

```
error[FLUX]: type cannot be refined
 --> crates/flux-tests/tests/todo/refined_fn_in_trait_01.rs:7:42
  |
7 | #[flux::sig(fn<T, refine q: T -> bool> (&T{v:q(v)}) -> T{v: q(v)})]
  |                                          ^
-Ztrack-diagnostics: created at crates/flux-desugar/src/desugar.rs:678:37
```

3. (In `main`) if you do write `T as base` then it works... ?

Reply:
The crash happens because the test is "wrong". The refined signature is

```rust
#[flux::sig(fn<refine q: T -> bool>(&T{v:q(v)}) -> T{v: q(v)})]
```
 but it should be 
 ```rust
#[flux::sig(fn<T, refine q: T -> bool>(&T{v:q(v)}) -> T{v: q(v)})]
```
is missing the `T`.

If you put the angle brackets in the refined signature you have to specify all the generics. We only auto-generate the generics if there are no angle brackets.

The crash should be fixed somewhere else.

Reply:
Hmm - still not following. 

Even if you write the `fn<T, refine...>` as you suggest, you *still* get the `T cannot be refined error`. 

So why not just cut to the chase and make it return that error in this case, right away? 

I mean sure you could have some other logic to check that all `T` are specified in the angle brackets but 

(a) That seems overkill  _why_ should I have to write _everything_ including unnecessary stuff (e.g. `T`) just because I wanted to write _one_ (e.g. `refine q: T -> bool`) thing, and 

(b) Anyways, the _real_ problem here is you need `T` to be refinable?

Reply:
I guess your point is 

1. we should just auto-generate the generics (to avoid the crash),
2. then the rest of the code would automatically point out the 'cannot refine' error?

Reply:
I think adding the "check-signature-has-enough-generics" is a bit pointless, for the reasons above. So really, we should probably do (1) which may be easy or hard (IDK) but at least doing the `Option` seems a strict improvement over the current crash?

Reply:
My point is that making `get_param` return `Option` is not fixing the underlying problem. If you revert that change and write `fn<T, refine...>` you still get the `T cannot be refined error`. 

The underlying problem is that we assume the refined signature has the same number of generics as the rust signature.

We can't avoid checking the list of generics in the refined signature matches the generics in the rust signature, e.g., the following should give an error (I think right now silently adds an extra generic)

```rust
#[flux::sig(fn<A, B>(x: A) )]
fn foo<A>(x: A) {}
```

The reason that generics can be specified in the refined signature is in case you want to specify they are of kind `base`

Reply:
Yes, it makes sense to write `<T as base,...>` my point is only that it doesn't make sense to ever (?) require the user to write `<T,...>` ...

Reply:
I fixed this in `desugar_generics` 
1. lifting the rust generics (to make sure we got all the type variables) and then
2. layering on the surface spec (computed by `desugar_surface_generics`) to mark variables as `base` if so specified.